libyul/optimiser/VariableAssignmentMap.h

```diff
@@ -0,0 +1,86 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#pragma once
+
+#include <libyul/YulString.h>
+
+#include <set>
+#include <unordered_map>
+
+namespace solidity::yul
+{
+
+/**
+ * Class that implements a reverse lookup for an ``unordered_map<YulString, set<YulString>>`` by wrapping the
+ * two such maps - one ordered, and one reversed, e.g.
+ *
+ *  m_ordered           m_reversed
+ *  f -> (g,)           g -> (f,)
+ *  c -> (b,d,e,)       d -> (c,)
+ *  a -> (b,c,)         c -> (a,)
+ *                      e -> (c,)
+ *                      b -> (a,c,)
+ *
+ * The above example will from here onwards be referenced as ```Ref 1```.
+ *
+ * This allows us to have simultaneously managed insertion and deletion via a single interface, instead of manually
+ * managing this at the point of usage (see ``DataFlowAnalyzer``).
+ */
+class VariableAssignmentMap
+{
+public:
+	VariableAssignmentMap() = default;
+	/**
+	 * Insert a set of values for the provided key ``_variable`` into ``m_ordered`` and ``m_reversed``.
+	 * This method will erase all references of ``_variable`` from both sets before performing the insertion,
+	 * practically making this an insert with overwrite method, akin to container assignment with subscript operator,
+	 * i.e. container[index] = value.
+	 * For example, if ``_variable`` is ``x`` and ``_references`` is ``{"y", "z"}``, the following would be added to ``Ref 1``:
+	 *
+	 * m_ordered        m_reversed
+	 * x -> (y, z,)     y -> (x,)
+	 *                  y -> (z,)
+	 *
+	 * @param _variable current expression variable
+	 * @param _references all referenced variables in the expression assigned to ``_variable``
+	 */
+	void insert(YulString const& _variable, std::set<YulString> const& _references);
+
+	/**
+	 * Erase entries in both maps based on provided ``_variable``.
+	 * For example, after deleting ``c`` for ``Ref 1``, ``Ref 1`` would contain the following:
```

Comment:
Can we say that `erase(x)` is exactly the same as `set(x, {})`?

Reply:
Eh, yes and no - since we do `m_ordered[_variable] = _references;`, which will still make an insertion for key `x`. I don't think it would ultimately affect the behaviour. I could insert an empty check for `_references` however, in which case your statement would be fully correct.

Reply:
The edit's the final answer - i.e. yes, it's exactly as Chris suggested; adding the empty `_references` check would alter the behaviour of the analysis (I would assume we'd see failing tests, but I'd have to check). I.e. we fetch by key, and then use the value set to either perform arithmetic (i.e. add two sets together), or lookup, neither of which need an empty  set check.

Reply:
ok, you're right about this altering the analysis.

But I'm still confused as to why you think `erase(x)` and `set(x, {})` would be equivalent. This just does not seem true to me. Are you referring to the fact that `m_ordered[_variable]` will modify `m_ordered` and insert the key if it's not there? You're still doing `m_ordered.erase()` at the end of the function so yeah, it will technically insert the key but the key won't be there when the function finishes. So I don't think it's true that ```The behaviour is the same as ``set("x", {})``.```. This bit should be removed from the docstring unless I'm missing something here.